src/finn/util/exception.py

```diff
@@ -41,3 +58,113 @@ class FINNDataflowError(FINNInternalError):
+        @functools.wraps(step)
+        def wrapped(model: ModelWrapper, cfg: DataflowBuildConfig) -> ModelWrapper:
+            try:
+                return step(model, cfg)
+            except Exception as e:
+                date = datetime.today().strftime("%d-%m-%Y__%I-%M-%S")
+                if build_dir_prefix is None:
+                    prefix = f"crash_{date}_"
+                else:
+                    prefix = f"{build_dir_prefix}_{date}_"
+                path = Path(make_build_dir(prefix))
+                if snapshot_model:
+                    # Get the frame where the exception was raised, get it's frame object,
+                    # and from it all locals, hopefully containing our ModelWrapper object
+                    error_locals = inspect.trace()[-1][0].f_locals
+                    modelwrappers = {
+                        k: v for k, v in error_locals.items() if isinstance(v, ModelWrapper)
+                    }
+                    actual_model: ModelWrapper | None = None
+                    if "model" in modelwrappers.keys():
+                        actual_model = modelwrappers["model"]
+                        actual_model.save(str(path / "transformation_model_snapshot.onnx"))
+                    elif len(modelwrappers.keys()) > 0:
+                        actual_model = next(iter(modelwrappers.values()))
+                        actual_model.save(str(path / "likely_transformation_model_snapshot.onnx"))
+                    else:
+                        actual_model = model
+                        model.save(str(path / "before_failed_transformation_model_snapshot.onnx"))
+                    if actual_model is not None:
+                        for node in model.graph.node:
+                            if node.op_type == "StreamingDataflowPartition":
+                                submodel = Path(getCustomOp(node).get_nodeattr("model"))
+                                if submodel.exists():
+                                    submodel_dir = Path(path / "submodels")
+                                    submodel_dir.mkdir()
+                                    shutil.copy(submodel, submodel_dir)
+                if snapshot_config:
+                    yaml = str(cfg.to_yaml())
+                    with (path / "cfg.yaml").open("w+") as f:
+                        f.write(yaml)
+                if snapshot_finn:
+                    finn_root = Path(finn.__file__).parent
+                    shutil.copytree(finn_root, path / "finn")
+                if snapshot_buildlog and cfg.output_dir is not None:
+                    shutil.copy(
+                        Path(cfg.output_dir) / "build_dataflow.log", path / "build_dataflow.log"
+                    )
+                    with (path / "build_dataflow.log").open("a") as f:
+                        f.write(traceback.format_exc())
+                if snapshot_finn_envvars:
+                    env: dict[str, str] = {}
+                    for key in [
```

Comment:
Is this list complete? How do we ensure this list will be updated when adding new `envvars` in the future - this is probably more a general concern: Seems that except for `NUM_DEFAULT_WORKERS` all have the `FINN_` prefix, does it make sense to save all FINN-prefixed environment variables (in a addition to some manually specified ones)?